### PR TITLE
Implement Jigsaw task

### PR DESCRIPTION
**Pull request by RPC2**
This PR aims to finish the implementation and tuning of the Jigsaw task.

### CODE REVIEW

File: src/il_representations/algos/__init__.py [hunk 1]
```diff
@@ -272,25 +275,32 @@ def __init__(self, **kwargs):
...
         encoder_kwargs = kwargs.get('encoder_kwargs') or {}
         encoder_cls_key = encoder_kwargs.get('obs_encoder_cls', None)
-
-        algo_hardcoded_kwargs = dict(encoder=VAEEncoder,
-                                     decoder=PixelDecoder,
+        _this_file_dir = os.path.dirname(os.path.abspath(__file__))
+        data_root = os.path.abspath(os.path.join(_this_file_dir, '../../../'))
+        permutation_filename = 'jigsaw_permutations_1000.npy'
+        permutation_file = \
+            glob.glob(f'{data_root}/**/{permutation_filename}', recursive=True)[0]
```
Comment (qxcv) [inline src/il_representations/algos/__init__.py:289]: Is there a reason to have these stored in a file? If the aim is just to obtain determinism, then we could use something like this:

```python
import numpy as np

permutations_seed = 42  # change to get different set of permutations
n_permutations = 1000
n_tiles = 9

assert n_permutations < np.math.factorial(n_tiles), \
    f"cannot get {n_permutations} permutations with only {n_tiles} tiles"
rng = np.random.RandomState(permutations_seed)
unique_permutations = set()
while len(unique_permutations) < n_permutations:
    # We iteratively add permutations to a set in order to avoid duplicates.
    # With n_tiles=9, n_permutations=1000 we are highly likely to
    # get at least one duplicate
    new_permutation = tuple(rng.permutation(n_tiles))
    # uncomment this to see duplicates
    # if new_permutation in unique_permutations: print('dupe!')
    unique_permutations.add(new_permutation)
permutations = np.asarray(sorted(unique_permutations))
# shuffle the permutations so there's less structure in class IDs
permutations = permutations[rng.permutation(len(permutations))]
```

File: src/il_representations/algos/pair_constructors.py [hunk 2]
```diff
@@ -167,3 +170,81 @@ def __call__(self, data_iter):
...
+
+        # We -2 after division here because, as the original paper indicated,
+        # preserving the whole image tile may let the network exploit edge information
+        # and not really learn useful representations. Therefore, here we are using
+        # the central-cropped version of these image tiles.
+        tile_h, tile_w = int(original_h / math.sqrt(self.n_tiles)) - 2, \
+                         int(original_w / math.sqrt(self.n_tiles)) - 2
+
+        permutation_class = [x for x in range(len(self.permutation))]
+        random.shuffle(permutation_class)
```
Comment (qxcv) [inline src/il_representations/algos/pair_constructors.py:198]:
```suggestion
        permutation_class = np.random.permutation(len(self.permutation))
```

File: src/il_representations/algos/pair_constructors.py [hunk 3]
```diff
@@ -167,3 +170,81 @@ def __call__(self, data_iter):
...
+            processed_obs_list = []
+            target_list = []
+            for o in obs:
+                processed_obs = self.make_jigsaw_puzzle(o,
+                                                        permutation_class[permute_idx],
+                                                        tile_h,
+                                                        tile_w)
+                processed_obs_list.append(processed_obs)
+            target_list.append(permutation_class[permute_idx])
+            permute_idx = (permute_idx + 1) % len(self.permutation)
```
Comment (qxcv) [inline src/il_representations/algos/pair_constructors.py:213]: Why do label assignment cyclically? You could just choose a random permutation class each time instead. That way there are no correlations between labels across batches.

```suggestion
            target_list.append(np.random.randint(len(self.permutation)))
```

File: src/il_representations/algos/pair_constructors.py [hunk 4]
```diff
@@ -167,3 +170,81 @@ def __call__(self, data_iter):
...
+        img_tiles = []
+        unit_h = int(image.shape[0] / n_tiles_sqrt)
+        unit_w = int(image.shape[1] / n_tiles_sqrt)
+
+        for pos in permute:
+            pos_h = int(pos // n_tiles_sqrt) * unit_h
+            pos_w = int(pos % n_tiles_sqrt) * unit_w
+
+            # We +1 after pos_h and pos_w to center crop
+            processed_image = image[pos_h + 1:pos_h + 1 + tile_h, pos_w + 1:pos_w + 1 + tile_w]
```
Comment (qxcv) [inline src/il_representations/algos/pair_constructors.py:242]: Is the image NCHW or NHWC? This code will only work in the latter case, so I suggest adding an assert to make sure that is the case. I also suggest adding an assert to make sure that `processed_image` has dimensions `tile_h*tile_w*channels`.

File: src/il_representations/scripts/il_train.py [hunk 5]
```diff
@@ -189,6 +189,22 @@ def make_policy(observation_space,
         else:
             encoder = encoder_or_path
         assert isinstance(encoder, nn.Module)
+
+        # Sometimes we are not using FC layers in encoders, e.g., Jigsaw.
+        # Hence in il training we need to recover this part of the encoder.
```
Comment (qxcv) [inline src/il_representations/scripts/il_train.py:194]: Is it possible to have some more information here? I'm not totally sure what this code does, or why it's needed.

**Review by qxcv — CHANGES_REQUESTED** [review]
Wow, this PR has been around for a while! At the moment I'm going through our PRs and trying to get everything merged into master. This branch is the base for a few other branches, so I'm trying to merge it first. Could you please have a look at these comments when you get the time, @RPC2?

File: src/il_representations/algos/__init__.py [hunk 6]
```diff
@@ -272,25 +275,32 @@ def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
 
-class Autoencoder(RepresentationLearner):
```
[the remark this answers is not in the document]
Comment (RPC2) [reply src/il_representations/algos/__init__.py]: Hmm it's pretty weird that none of my commits did this but they collectively removed this class. I've added it back. Thanks for catching that!

File: src/il_representations/algos/decoders.py [hunk 7]
```diff
@@ -68,7 +68,10 @@ def get_sequential_from_architecture(architecture, representation_dim, projectio
         layers.append(nn.BatchNorm1d(num_features=layer_def['output_dim']))
         input_dim = layer_def['output_dim']
     layers.append(nn.Linear(input_dim, projection_dim))
-    return nn.Sequential(*layers)
+    network = nn.Sequential(*layers)
+    if torch.cuda.is_available():
+        network = network.to(torch.device('cuda'))
```
[the remark this answers is not in the document]
Comment (RPC2) [reply src/il_representations/algos/decoders.py]: Fixed that!

File: src/il_representations/algos/__init__.py (hunk 1 above)
Comment (RPC2) [reply src/il_representations/algos/__init__.py:313]: Since the original Jigsaw paper requires that the 1000 permutations be sufficiently different (i.e., have a large hamming distance among each other), generating these permutations requires calculating the hamming distance once per iteration, which can take a long time as the size of the list of selected permutations grows (e.g., it takes 40 minutes on perceptron). So I just saved this list as a file and reload it every time. I have added a function into `utils.py` if people want to generate it again.

```
    all_perms = np.array(list(itertools.permutations(list(range(n_tiles)), n_tiles)))

    perms = []
    for i in range(n_perms):
        if i == 0:
            j = np.random.randint(len(all_perms))
            perms = np.array(all_perms[j]).reshape([1,-1])
        else:
            perms = np.concatenate([perms, all_perms[j].reshape([1,-1])], axis=0)

        all_perms = np.delete(all_perms, j, axis=0)
        
        # Calculate selected permutations' distance with all possible
        # permutations, and select the one that are most different from
        # selected ones. This is to make sure the generated jigsaw puzzles are
        # sufficiently different from each other.
        # This takes a long time when len(perms) grows.
        D = cdist(perms, all_perms, metric='hamming').mean(axis=0).flatten()
        j = D.argmax()
```

File: src/il_representations/algos/encoders.py [hunk 8]
```diff
@@ -525,6 +539,52 @@ def encode_target(self, x, traj_info):
...
+        Return:
+            img_tiles (tensor) - Processed image tiles with shape [batch_size, 9, C, H/3, W/3]
+        """
+        batch_size, c, h, w = img.shape
+        unit_size = math.sqrt(self.n_tiles)
+
+        h_unit = int(h/unit_size)
+        w_unit = int(w/unit_size)
+
+        return img.view(batch_size, self.n_tiles, c, h_unit, w_unit)
```
[the remark this answers is not in the document]
Comment (RPC2) [reply src/il_representations/algos/encoders.py]: Fixed it in later commits :).

File: src/il_representations/algos/encoders.py [hunk 9]
```diff
@@ -525,6 +539,52 @@ def encode_target(self, x, traj_info):
...
+
+    def img_to_tiles(self, img):
+        """
+        Input:
+            img (tensor) - image to be processed, with shape [batch_size, C, H, W]
+        Return:
+            img_tiles (tensor) - Processed image tiles with shape [batch_size, 9, C, H/3, W/3]
+        """
+        batch_size, c, h, w = img.shape
+        unit_size = math.sqrt(self.n_tiles)
```
[the remark this answers is not in the document]
Comment (RPC2) [reply src/il_representations/algos/encoders.py:578]: I've added a few lines so that if `n_tiles` is not a square number, the class's `__init__` function will raise an error. Also, if `h` or `w` are not multiples of `unit_size`, the program will raise a warning.

File: src/il_representations/scripts/il_train.py (hunk 5 above)
Comment (RPC2) [reply src/il_representations/scripts/il_train.py]: Added more comments just now.

File: src/il_representations/algos/__init__.py (hunk 1 above)
Comment (qxcv) [reply src/il_representations/algos/__init__.py:331]: Okay, got it. Thanks!

File: src/il_representations/scripts/il_train.py [hunk 10]
```diff
@@ -190,15 +190,22 @@ def make_policy(observation_space,
             encoder = encoder_or_path
         assert isinstance(encoder, nn.Module)
 
-        # Sometimes we are not using FC layers in encoders, e.g., Jigsaw.
-        # Hence in il training we need to recover this part of the encoder.
         encoder_net = encoder.network if 'network' in dir(encoder) else \
             encoder.query_encoder.network
+
+        # Normally the last layer of an encoder is a linear layer, but in
+        # some special cases like Jigsaw, we only train the convolution
```
Comment (qxcv) [inline src/il_representations/scripts/il_train.py:292]: Thanks for adding this explanation! Why does Jigsaw need to push the linear layer into the decoder, instead of leaving it in the encoder?

**Review by qxcv — COMMENTED** [review]
Thanks for making those changes! Two more questions about network structure then I think we're good to merge.

File: src/il_representations/algos/decoders.py [hunk 11]
```diff
@@ -276,10 +274,12 @@ def forward(self, z, traj_info, extra_context=None):
         # For Jigsaw, the z input dimension depends on state observation shape;
```
[the remark this answers is not in the document]
Comment (RPC2) [reply src/il_representations/algos/decoders.py:271]: This is because Jigsaw has a different way to pass an image. Its forward pass is like this (assuming cutting an image into 3x3 tiles):

```
tiles = image_to_tiles(images)  # [B, C, H, W] -> [B, 9, C, H/3, W/3]

encoder_output = []
for i in range(9):
    tile = x[:, i, :, :, :]
    encoder_representation = self.forward(tile)   # [B, dim]
    encoder_output.append(encoder_representation)
encoder_output = torch.cat(encoder_output, axis=1)  # [B, dim*9]
...
decoder_output = self.decoder(encoder_output)
```

The `encoder_output` is obtained by concatenating 9 `encoder_representation` generated by the encoder network. This is different from other algorithms, where we obtain `encoder_output` just from one network forward pass. This can induce several changes:

1. With the same convolution layers, image tiles with [C, H/3, W/3] produce a different output dimension from an image with shape [C, H, W]. This means if we add a linear layer after conv layers in the encoder during RepL, this linear layer can't be reused for IL. Example:
```
# In encoder.py
t = torch.rand([64, 9, 31, 31]).cuda()
self.forward(t, traj_info=None)
>>> Independent(Normal(loc: torch.Size([64, 128]), scale: torch.Size([64, 128])), 1)

t = torch.rand([64, 9, 100, 100]).cuda()   # DMC tasks
self.forward(t, traj_info=None)
>>>Independent(Normal(loc: torch.Size([64, 1568]), scale: torch.Size([64, 1568])), 1)
```
So instead of having a linear layer in the encoder and discard it during IL, I just directly set the encoder to contain conv layers only, and let the decoder handle the linearity.

2. Since the `encoder_output` shape eventually depends on `observation_space` and `encoder.network`'s structure instead of a preset `representation_dim`, we either set the decoder `projection_layers` by passing in `observation_space` and `encoder.network` as a kwarg during initialization or adjust it upon receiving an input. I think adjusting it on the fly will make fewer changes to the code, so I wrote the code block starting L274 above. To avoid constantly changing `self.projection_layers`, I just added an assertion so it is only allowed to change once (i.e. when it first receives an input).

File: src/il_representations/scripts/il_train.py (hunk 10 above)
Comment (RPC2) [reply src/il_representations/scripts/il_train.py:293]: I explained these all together in the comment above :).

File: src/il_representations/algos/decoders.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (qxcv) [reply src/il_representations/algos/decoders.py:271]: Thanks for the thorough explanation! I looked at fig 3 of the paper & it seems that this is how they do it too. The main thing I'm concerned about is someone saving the model, reloading it, evaluating on a different dataset, and having it silently reset its internal weights. Now that we have the assertion I think we should be pretty good; I might add a warning too so that the user can see what is going on (& decide whether it is desirable).

**Review by qxcv — APPROVED** [review]
marking approved
